packages/gatsby-cli/src/create-cli.js

```diff
@@ -71,8 +73,28 @@ function buildLocalCommands(cli, isLocalSite) {
           `There was a problem loading the local ${command} command. Gatsby may not be installed in your site's "node_modules" directory. Perhaps you need to run "npm install"? You might need to delete your "package-lock.json" as well.`
         )
 
-      report.verbose(`loading local command from: ${cmdPath}`)
-      return require(cmdPath)
+      if (command === `develop`) {
+        return args => {
+          const tmpFile = require(`tmp`).fileSync()
+          require(`fs`).writeFileSync(
+            tmpFile.name,
+            `const cmd = require("${cmdPath}");
+const args = ${JSON.stringify(args)};
+cmd(args);`
+          )
+          const monitor = respawn([`node`, tmpFile.name], {
+            env: process.env,
+          })
+          chokidar
+            .watch(path.join(process.cwd(), `gatsby-config.js`))
```

Comment:
Check out https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/bootstrap/index.js#L225

Those are files that we probably should also watch for. Not 100% sure about watching node_modules, but at least the rest of the gatsby config/framework files.

Reply:
I don't think we want to watch `package.json`, that seems excessive, however good call on the `gatsby-node.js`!

Reply:
Why not? We should restart when new packages are installed I think.

Reply:
Also I can easily imagine admin adding new dependencies.

Reply:
I don't think that's necessary? When I `yarn add redux` and then start using redux on my Gatsby site, the develop process doesn't need to be restarted from scratch, it'll just work, right?

Reply:
Well, if you eg upgrade a plugin version you def need a restart.

Reply:
Yeah that's an interesting edge case. I feel like restarting the dev process for every package.json change would be more annoying than helpful, however we could look into adding smarter heuristics specific to package.json in the future?

Reply:
Sure, but I think currently it's better to restart server than not to.

Reply:
> I think currently it's better to restart server than not to.

I disagree, here's why: I would be very annoyed if I had to wait 5-10 seconds for the develop process to bootstrap again before I can work on my website every time I add a dependency. That is much worse than right now, where I can just keep working on my website as if nothing happened.

So in my mind, we are making a common scenario (adding a dependency and other non-breaking changes to the pkg.json like editing the name) a much worse experience for our users, for a _slightly_ better experience with a less common edge case (upgrading a dependency).

I would prefer we keep the experience the same overall and then make it better over time, not make some (more common) scenarios worse to make some others slightly better?

Reply:
Maybe add a manual restart option?

Reply:
There are more problems with watching `package.json`:
 - user can manually edit `package.json` and not install deps yet - so restarting dev server on manual edits just won't do anything.

We could watch `package.json` and on changes try to re-construct final config object (which contains package versions of plugins) and compare current config to new one and restart if those are different. This would prevent restarts when adding `redux`, but would still cause restarts when up/downgrading plugins?

Reply:
That sounds like a great soultion @pieh

Reply:
Yep exactly @pieh! 👍  That's what I meant with smarter heuristics above.